### PR TITLE
 [AF-913]: Added to CommandBuilder new methods with new Elemental2 Elements

### PR DESCRIPTION
**Pull request by adrielparedes**
Added categories to ResourceTypeDefinitions to filter assets.

This is related to new UX Assets design.

Issues related:
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/686
https://github.com/kiegroup/appformer/pull/116
https://github.com/kiegroup/kie-wb-common/pull/1317
https://github.com/kiegroup/drools-wb/pull/727
https://github.com/kiegroup/jbpm-designer/pull/732
https://github.com/kiegroup/jbpm-wb/pull/966
https://github.com/kiegroup/kie-wb-distributions/pull/664

### CODE REVIEW

File: uberfire-project/uberfire-project-client/src/main/java/org/guvnor/common/services/project/client/type/POMResourceType.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 import com.google.gwt.user.client.ui.IsWidget;
 import org.guvnor.common.services.project.editor.type.POMResourceTypeDefinition;
 import org.uberfire.client.workbench.type.ClientResourceType;
+import org.uberfire.workbench.type.Category;
```
Comment (jomarko) [inline uberfire-project/uberfire-project-client/src/main/java/org/guvnor/common/services/project/client/type/POMResourceType.java]: Unused import, did you plan to override `getCategory()`?

File: uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package org.uberfire.workbench.type;
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java]: License please.

File: uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+public enum Category {
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java]: We've previously tried to keep AppFormer clean of KIE related terms (IDK if this still happens!?!?) as AppFormer was meant to be a generic _workbench_ framework and not specific for KIE. Therefore I question whether we should have this ```enum``` in AppFormer. At best have some abstraction that is implemented in ```kie-wb-common```?

It is however a while since I was directly involved with AppFormer so I could be out of date.. hence happy to defer the decision to @ederign or @mbarkley

Comment (mbarkley) [reply uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java]: Thanks for tagging me in. New plan is that it doesn't make sense to split AppFormer from the rest of our platform artificially when the reality is that we mostly care about KIE-related use cases.

So KIE terminology is fine and after we fix the build, AppFormer will depend on droolsjbpm-knowledge. AppFormer should not depend on drools, jbpm, or optaplanner directly though.

File: uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java [hunk 4]
```diff
@@ -68,4 +68,12 @@
      * @return true if matches, otherwise false
      */
     boolean accept(final Path path);
+
+    /**
+     * Defines the category for each resource. If no category is defined a default is set
+     * @return the resource category
+     */
+    default Category getCategory() {
+        return Category.MORE;
```
Comment (paulovmr) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java]: Personally I think `OTHERS` would be a better word to that category.

**Comment by porcelli** [comment]
no tests for select UI?

File: uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.uberfire.workbench.type;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public enum Category {
+
+    OTHERS,
```
Comment (ederign) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java]: In the ideal world, Category on uberfire api doesn't know anything about the asset types.

My suggestion is to extract this and somehow dinamycally lookup the categories. One option is to use cdi for this, like is done with SocialAdapter and his child classes.

Comment (ederign) [reply uberfire-api/src/main/java/org/uberfire/workbench/type/Category.java]: another option would be to transform category in a cdi bean, scan all resource type instances (they are all beans) and dynamically lookup for all category types.

Categories are only used by resource types?

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Category.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.workbench.category;
+
+public interface Category {
+
+//    OTHERS,
```
Comment (ederign) [inline uberfire-api/src/main/java/org/uberfire/workbench/category/Category.java]: ?

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.uberfire.ext.wires.bpmn.api.category;
+
+import org.uberfire.workbench.category.Category;
+
+public final class Process implements Category {
+
+    @Override
+    public String getName() {
+        return "PROCESS";
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java]: Could you please extract this for constants in the class?

File: uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java [hunk 8]
```diff
@@ -68,4 +70,12 @@
      * @return true if matches, otherwise false
      */
     boolean accept(final Path path);
+
+    /**
+     * Defines the category for each resource. If no category is defined a default is set
+     * @return the resource category
+     */
+    default Category getCategory() {
+        return new Others();
```
Comment (ederign) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java]: This is at least risky. Others is a managed bean and we should never create new instances.

Comment (adrielparedes) [reply uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java]: @ederign so I couldn't do this in there? That was the solution to not include the Others category in every class, so It was going to be the default Category for every ResourceType

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-api/src/main/java/org/uberfire/ext/plugin/type/SplashPluginResourceTypeDefinition.java [hunk 9]
```diff
@@ -35,4 +46,9 @@ public String getDescription() {
     public String getSuffix() {
         return "/" + PluginType.SPLASH.toString().toLowerCase() + ".plugin";
     }
+
+    @Override
+    public Category getCategory() {
+        return null;
```
Comment (jhrcek) [inline uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-api/src/main/java/org/uberfire/ext/plugin/type/SplashPluginResourceTypeDefinition.java]: What's the purpose of the field `private Category category` and initializing it in the constructor, when you're always returning null here?

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ */
+
+package org.uberfire.ext.wires.bpmn.api.category;
+
+import org.uberfire.workbench.category.Category;
+
+public final class Process extends Category {
+
+    public static final String CATEGORY_NAME = "PROCESS";
```
Comment (jhrcek) [inline uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java]: Is it useful to have this public? The same applies to other category subclasses throughout this PR.

File: uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java [hunk 11]
```diff
@@ -68,4 +70,10 @@
      * @return true if matches, otherwise false
      */
     boolean accept(final Path path);
+
+    /**
+     * Defines the category for each resource. If no category is defined a default is set
```
Comment (jhrcek) [inline uberfire-api/src/main/java/org/uberfire/workbench/type/ResourceTypeDefinition.java]: Please add a comment about what category would be the default one if not defined..

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/select/SelectComponent.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @PostConstruct
+    public void init() {
+        this.view.init(this);
+    }
+
+    public void setup(List<SelectOption> options,
+                      Consumer<SelectOption> selectCallback) {
+        this.setSelected(options.get(0));
```
Comment (jhrcek) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/select/SelectComponent.java]: Is this intended to be generic SelectComponent? If yes and somebody provides empty list of options, you're getting `ArrayIndexOutOfBoundsException`  here ..

Also a test with empty options list would be nice.

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package org.guvnor.common.services.project.categories;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.uberfire.workbench.category.Category;
+
+@ApplicationScoped
+public class Model extends Category {
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java]: The name Model sounds too generic. Java/Data/Process/Form some other model? Could you make the category name a bit more specific?

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-api/src/main/java/org/uberfire/ext/plugin/type/SplashPluginResourceTypeDefinition.java (hunk 9 above)
Comment (adrielparedes) [reply uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-api/src/main/java/org/uberfire/ext/plugin/type/SplashPluginResourceTypeDefinition.java]: @jhrcek my mistake!

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Category.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+public abstract class Category {
+
+    public abstract String getName();
+
+    @Override
+    public boolean equals(Object category) {
+
+        if (category instanceof Category && category != null) {
+            Category cat = (Category) category;
+            if (cat != null && !cat.getName().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (adrielparedes) [reply uberfire-api/src/main/java/org/uberfire/workbench/category/Category.java]: @jhrcek thanks!

File: uberfire-api/src/main/java/org/uberfire/workbench/type/AnyResourceTypeDefinition.java [hunk 15]
```diff
@@ -17,9 +17,20 @@
...
 
 import org.uberfire.backend.vfs.Path;
+import org.uberfire.workbench.category.Category;
+import org.uberfire.workbench.category.Others;
 
 public class AnyResourceTypeDefinition implements ResourceTypeDefinition {
 
+    private Category category;
+
+    public AnyResourceTypeDefinition() {
```
[the remark this answers is not in the document]
Comment (adrielparedes) [reply uberfire-api/src/main/java/org/uberfire/workbench/type/AnyResourceTypeDefinition.java]: This classes need to be injected because are part of CDI and I have to find those classes to cache them. Also the categories needs to be "plug and play", so you need to find every class injecting Any<Category>. That was the initial premise to build this part. They can't be null as you said but they have to be injected.

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java (hunk 13 above)
Comment (adrielparedes) [reply uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java]: The category names were previously selected. And the resource type for those categories too. I can  give you the document where they are voted

**Comment by ederign** [comment]
please jenkins execute full downstream build

**Comment by ederign** [comment]
@adrielparedes this is the PR's right?

**Comment by adrielparedes** [comment]
@ederign, this contains categories and the caches
and this uses them: https://github.com/kiegroup/kie-wb-common/pull/1317

**Comment by adrielparedes** [comment]
@ederign tests failing in downstream build are not related, should I run it again?

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.ext.wires.bpmn.api.category;
+
+import org.uberfire.workbench.category.Category;
+
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java]: @ApplicationScoped ?

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/type/BpmnResourceTypeDefinition.java [hunk 17]
```diff
@@ -16,10 +16,27 @@
...
 
 public class BpmnResourceTypeDefinition implements ResourceTypeDefinition {
 
+    private Category category;
+
+    public BpmnResourceTypeDefinition() {
+
+    }
+
+    public BpmnResourceTypeDefinition(final Process category) {
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/type/BpmnResourceTypeDefinition.java]: This works?

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 18]
```diff
@@ -63,15 +62,31 @@
...
      */
     private final List<SplashScreenActivity> splashActivities = new ArrayList<SplashScreenActivity>();
-    @Inject
     private SyncBeanManager iocManager;
-    @Inject
     private Event<NewPerspectiveEvent> newPerspectiveEventEvent;
-    @Inject
     private Event<NewWorkbenchScreenEvent> newWorkbenchScreenEventEvent;
+    private CategoriesManagerCache categoriesManagerCache;
+    private ResourceTypeManagerCache resourceTypeManagerCache;
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: @adrielparedes why keep two instances of ActivityAndMetaInfo? I would delete 
private final List<ActivityAndMetaInfo> resourceActivities = new ArrayList<ActivityAndMetaInfo>();

and keep all the calls inside the resourceTypeManagerCache

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 19]
```diff
@@ -63,15 +62,31 @@
      * All active activities that are splash screens.
      */
     private final List<SplashScreenActivity> splashActivities = new ArrayList<SplashScreenActivity>();
-    @Inject
     private SyncBeanManager iocManager;
-    @Inject
     private Event<NewPerspectiveEvent> newPerspectiveEventEvent;
-    @Inject
     private Event<NewWorkbenchScreenEvent> newWorkbenchScreenEventEvent;
+    private CategoriesManagerCache categoriesManagerCache;
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: I would add some encapsulation here moving categories cache as a field of ResourceTypeManagerCache

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 20]
```diff
@@ -88,16 +103,36 @@ void init() {
             } else {
                 final Pair<Integer, List<String>> metaInfo = generateActivityMetaInfo(activityBean);
                 if (metaInfo != null) {
-                    getResourceActivities().add(new ActivityAndMetaInfo(activityBean,
-                                                                        metaInfo.getK1(),
-                                                                        metaInfo.getK2()));
+                    ActivityAndMetaInfo activityAndMetaInfo = new ActivityAndMetaInfo(activityBean,
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: I would move this for ResourceTypeManagerCache

Comment (ederign) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: also L114

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 21]
```diff
@@ -88,16 +103,36 @@ void init() {
...
+                                                                                      metaInfo.getK2());
+                    addResourceActivity(activityAndMetaInfo);
                 }
             }
         }
 
         sortResourceActivitiesByPriority();
     }
 
+    private void addResourceActivity(ActivityAndMetaInfo activityAndMetaInfo) {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: I would move this for ResourceTypeManagerCache

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 22]
```diff
@@ -88,16 +103,36 @@ void init() {
...
+        getResourceActivities().add(activityAndMetaInfo);
+        List<ClientResourceType> resourceTypes = getResourceTypes(activityAndMetaInfo);
+        getResourceTypeManagerCache().addAll(resourceTypes);
+        getCategoriesManagerCache().addAllFromResourceTypes(resourceTypes);
+    }
+
+    protected CategoriesManagerCache getCategoriesManagerCache() {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: this also.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java [hunk 23]
```diff
@@ -191,9 +221,9 @@ public void addNewEditorActivity(final SyncBeanDef<Activity> activityBean,
         activitiesById.put(id,
                            activityBean);
 
-        resourceActivities.add(new ActivityAndMetaInfo(activityBean,
-                                                       Integer.valueOf(priority),
-                                                       Arrays.asList(resourceTypeName)));
+        this.addResourceActivity(new ActivityAndMetaInfo(activityBean,
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/ActivityBeansCache.java]: I would also move this for cache.addNewEditor

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/type/BpmnResourceTypeDefinition.java (hunk 17 above)
Comment (adrielparedes) [reply uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/type/BpmnResourceTypeDefinition.java]: The important there is the subclass that has the injections and it is the only class that uses it.

**Comment by ederign** [comment]
please jenkins test this

**Comment by ederign** [comment]
Looks good, needs to be merged with: https://github.com/kiegroup/appformer/pull/142

**Comment by ederign** [comment]
please jenkins retest this

**Comment by adrielparedes** [comment]
jenkins execute full downstream build

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.uberfire.workbench.category;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class Others extends Category {
+
+    private static final String OTHERS = "OTHERS";
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java]: What makes me think the first thing we're asked to do is i18n these category names!?

Fortunately you can ```@Inject TranslationService ts``` and use that.

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Undefined.java [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.uberfire.workbench.category;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class Undefined extends Category {
+
+    private static final String UNDEFINED = "UNDEFINED";
```
Comment (manstis) [inline uberfire-api/src/main/java/org/uberfire/workbench/category/Undefined.java]: i18n

File: uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+package org.uberfire.ext.wires.bpmn.api.category;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.uberfire.workbench.category.Category;
+
+@ApplicationScoped
+public class Process extends Category {
+
+    private static final String CATEGORY_NAME = "PROCESS";
```
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-bpmn/uberfire-wires-bpmn-api/src/main/java/org/uberfire/ext/wires/bpmn/api/category/Process.java]: i18n

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Decision.java [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+package org.guvnor.common.services.project.categories;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.uberfire.workbench.category.Category;
+
+@ApplicationScoped
+public class Decision extends Category {
+
+    private static final String CATEGORY_NAME = "DECISION";
```
Comment (manstis) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Decision.java]: i18n

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Form.java [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+package org.guvnor.common.services.project.categories;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.uberfire.workbench.category.Category;
+
+@ApplicationScoped
+public class Form extends Category {
+
+    private static final String CATEGORY_NAME = "FORM";
```
Comment (manstis) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Form.java]: i18n

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+
+    private static final String CATEGORY_NAME = "MODEL";
```
Comment (manstis) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Model.java]: i18n

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Process.java [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+package org.guvnor.common.services.project.categories;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.uberfire.workbench.category.Category;
+
+@ApplicationScoped
+public class Process extends Category {
+
+    private static final String CATEGORY_NAME = "PROCESS";
```
Comment (manstis) [inline uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/categories/Process.java]: i18n

File: uberfire-testing-utils/src/test/java/org/uberfire/promise/SyncPromisesTest.java [hunk 31]
```diff
@@ -0,0 +1,248 @@
+package org.uberfire.promise;
```
Comment (manstis) [inline uberfire-testing-utils/src/test/java/org/uberfire/promise/SyncPromisesTest.java]: License

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+    public void addAll(Collection<Category> category) {
+        this.categories.addAll(category);
+    }
+
+    public void addAllFromResourceTypes(List<ClientResourceType> clientResourceType) {
+        this.addAll(this.getCategories(clientResourceType));
+    }
+
+    private List<Category> getCategories(List<ClientResourceType> resourceTypes) {
+        return resourceTypes.stream()
```
Comment (manstis) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java]: So we can have multiple instances of the same ```Category``` here....

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public Category getCategory(String filterType) {
+        if (filterType == null) {
+            return undefinedCategory;
+        } else {
+            return this.getCategories()
+                    .stream()
+                    .filter(category -> category.getName().equals(filterType.toUpperCase()))
+                    .findFirst()
```
Comment (manstis) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java]: ... and here you then find the first. Why not just add a single instance of each ```Category``` to ```this.categories``` in the first place?

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java (hunk 24 above)
Comment (adrielparedes) [reply uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java]: @manstis this is not for i18n, I have a place for it. This is just a String for some internal stuffs, but they are not intended to be shown to the user

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java (hunk 33 above)
Comment (adrielparedes) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java]: @manstis because I'm filtering all the ResourceTypes, and we could have in many different resource types the same Category, this is just to get the first category that matches.

**Comment by adrielparedes** [comment]
jenkins execute full downstream build

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java (hunk 33 above)
Comment (manstis) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/CategoriesManagerCache.java]: @adrielparedes That was my whole point.. the ```List<Category> getCategories(List<ClientResourceType> resourceTypes)``` on L#60 currently returns repetitions of a ```Category``` (for example of you have 2 ```ClientResourceType``` and both have a category of ```OTHER``` you'll have 2 ```OTHER``` in the list). This leads you to have to ```findFirst()``` on L#73. If L#60 returned a ```Set<Category>``` that only held unique categories with no repetition you wouldn't need to find the first match later on.. 

Anyway, the code you have "works" and both the number of client resource types and categories is small so changing to set and removing the need for "find first" hardly brings any performance gain. So I'll stop finding fault :-)

File: uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java (hunk 24 above)
Comment (manstis) [reply uberfire-api/src/main/java/org/uberfire/workbench/category/Others.java]: @adrielparedes Thanks for the clarification.

**Comment by adrielparedes** [comment]
@manstis done!

**Comment by ederign** [comment]
jenkins execute full downstream build

**Comment by adrielparedes** [comment]
@ederign I think something is wrong in somewhere with kie-wb-distribution (I think is on master or something like that) because It breaks at the end. I saw many PRs doing the same:

**[ERROR] Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:start (start-container) on project kie-wb-tests-gui: Artifact [org.kie:kie-wb:war] is not a dependency of the project. -> [Help 1]**

kie-wb-common downstream build fail because of this too.

**Comment by adrielparedes** [comment]
jenkins execute full downstream build
